File: crates/sc-dav/src/ifheader.rs
//! RFC 4918 §10.4 `If` header — recursive-descent parser and evaluator.
//!
//! ```text
//! If           = 1*( No-tag-list | Tagged-list )
//! No-tag-list  = List
//! Tagged-list  = Resource-Tag 1*List
//! List         = "(" 1*Condition ")"
//! Condition    = ["Not"] ( State-token | "[" entity-tag "]" )
//! State-token  = Coded-URL
//! Coded-URL    = "<" absolute-URI ">"
//! Resource-Tag = "<" Simple-ref ">"
//! ```
//!
//! The grammar is only locally ambiguous: at the top level a `<` starts a
//! Resource-Tag, inside parentheses it starts a State-token. That single fact
//! is what the parser is built around.
//!
//! Parse failure is a 400, an unsatisfied header is a 412, and a write to a
//! locked resource without the matching token is a 423 — those three outcomes
//! are decided by the caller from [`IfHeader::evaluate`] plus
//! [`IfHeader::tokens`].

use crate::error::{DavError, DavResult};

#[derive(Clone, PartialEq, Eq, Debug)]
pub enum CondKind {
    StateToken(String),
    /// Entity tag, *without* surrounding quotes; `weak` records `W/`.
    ETag { value: String, weak: bool },
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Condition {
    pub not: bool,
    pub kind: CondKind,
}

/// One parenthesised list, together with the resource it was tagged for
/// (`None` for a No-tag-list, meaning "the request URI").
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct IfList {
    pub tag: Option<String>,
    pub conditions: Vec<Condition>,
}

#[derive(Clone, PartialEq, Eq, Debug, Default)]
pub struct IfHeader {
    pub lists: Vec<IfList>,
}

/// What the server knows about one resource, for evaluating conditions.
#[derive(Clone, Debug, Default)]
pub struct ResourceState {
    /// Active lock tokens on the resource, in `urn:uuid:…` form.
    pub tokens: Vec<String>,
    /// Current entity tag, unquoted.
    pub etag: Option<String>,
    /// False when the resource does not exist; no condition can match then
    /// except a negated one.
    pub exists: bool,
}

struct Cursor<'a> {
    b: &'a [u8],
    i: usize,
}

impl<'a> Cursor<'a> {
    fn peek(&self) -> Option<u8> {
        self.b.get(self.i).copied()
    }
    fn bump(&mut self) -> Option<u8> {
        let c = self.peek();
        if c.is_some() {
            self.i += 1;
        }
        c
    }
    fn skip_ws(&mut self) {
        while matches!(self.peek(), Some(c) if c == b' ' || c == b'\t' || c == b'\r' || c == b'\n')
        {
            self.i += 1;
        }
    }
    fn eof(&self) -> bool {
        self.i >= self.b.len()
    }
    /// Case-insensitive literal match that also refuses to swallow a longer
    /// token (`Nothing` must not parse as `Not` + `hing`).
    fn eat_keyword(&mut self, kw: &str) -> bool {
        let n = kw.len();
        if self.i + n > self.b.len() {
            return false;
        }
        if !self.b[self.i..self.i + n].eq_ignore_ascii_case(kw.as_bytes()) {
            return false;
        }
        match self.b.get(self.i + n) {
            None => {}
            Some(c) if c.is_ascii_whitespace() || *c == b'<' || *c == b'[' || *c == b'(' => {}
            Some(_) => return false,
        }
        self.i += n;
        true
    }
}

const MAX_LISTS: usize = 256;
const MAX_CONDS: usize = 256;
const MAX_TOKEN_LEN: usize = 2048;

impl IfHeader {
    pub fn parse(s: &str) -> DavResult<IfHeader> {
        let mut c = Cursor {
            b: s.as_bytes(),
            i: 0,
        };
        let mut lists = Vec::new();
        c.skip_ws();
        if c.eof() {
            return Err(DavError::BadRequest("empty If header".into()));
        }
        while !c.eof() {
            if lists.len() > MAX_LISTS {
                return Err(DavError::BadRequest("If header has too many lists".into()));
            }
            match c.peek() {
                Some(b'(') => {
                    let conds = parse_list(&mut c)?;
                    lists.push(IfList {
                        tag: None,
                        conditions: conds,
                    });
                }
                Some(b'<') => {
                    let tag = parse_coded_url(&mut c)?;
                    c.skip_ws();
                    // Tagged-list requires at least one List.
                    if c.peek() != Some(b'(') {
                        return Err(DavError::BadRequest(
                            "resource tag not followed by a list".into(),
                        ));
                    }
                    while c.peek() == Some(b'(') {
                        if lists.len() > MAX_LISTS {
                            return Err(DavError::BadRequest(
                                "If header has too many lists".into(),
                            ));
                        }
                        let conds = parse_list(&mut c)?;
                        lists.push(IfList {
                            tag: Some(tag.clone()),
                            conditions: conds,
                        });
                        c.skip_ws();
                    }
                }
                Some(other) => {
                    return Err(DavError::BadRequest(format!(
                        "unexpected byte {:?} in If header",
                        other as char
                    )))
                }
                None => break,
            }
            c.skip_ws();
        }
        if lists.is_empty() {
            return Err(DavError::BadRequest("If header has no lists".into()));
        }
        Ok(IfHeader { lists })
    }

    /// Every state token mentioned anywhere in the header, negated or not.
    /// The lock layer uses this to decide 423 vs. proceed.
    pub fn tokens(&self) -> Vec<&str> {
        let mut out = Vec::new();
        for l in &self.lists {
            for c in &l.conditions {
                if let CondKind::StateToken(t) = &c.kind {
                    if !c.not {
                        out.push(t.as_str());
                    }
                }
            }
        }
        out
    }

    /// RFC 4918 §10.4.3: lists are OR-ed, conditions inside a list are AND-ed.
    ///
    /// `state_of` is asked for the state of a tagged resource (`Some(url)`) or
    /// of the request URI (`None`).
    pub fn evaluate<F>(&self, mut state_of: F) -> bool
    where
        F: FnMut(Option<&str>) -> ResourceState,
    {
        for list in &self.lists {
            let st = state_of(list.tag.as_deref());
            let mut all = true;
            for cond in &list.conditions {
                let hit = match &cond.kind {
                    CondKind::StateToken(t) => st.tokens.iter().any(|x| x == t),
                    CondKind::ETag { value, weak: _ } => {
                        st.exists && st.etag.as_deref() == Some(value.as_str())
                    }
                };
                if hit == cond.not {
                    all = false;
                    break;
                }
            }
            if all {
                return true;
            }
        }
        false
    }
}

fn parse_list(c: &mut Cursor<'_>) -> DavResult<Vec<Condition>> {
    debug_assert_eq!(c.peek(), Some(b'('));
    c.bump();
    let mut conds = Vec::new();
    loop {
        c.skip_ws();
        match c.peek() {
            Some(b')') => {
                c.bump();
                break;
            }
            None => return Err(DavError::BadRequest("unterminated If list".into())),
            _ => {}
        }
        if conds.len() > MAX_CONDS {
            return Err(DavError::BadRequest("too many conditions".into()));
        }
        let not = c.eat_keyword("Not");
        c.skip_ws();
        let kind = match c.peek() {
            Some(b'<') => CondKind::StateToken(parse_coded_url(c)?),
            Some(b'[') => parse_etag(c)?,
            Some(other) => {
                return Err(DavError::BadRequest(format!(
                    "unexpected byte {:?} in If condition",
                    other as char
                )))
            }
            None => return Err(DavError::BadRequest("truncated If condition".into())),
        };
        conds.push(Condition { not, kind });
    }
    if conds.is_empty() {
        // `List = "(" 1*Condition ")"` — an empty list is a syntax error.
        return Err(DavError::BadRequest("empty If list".into()));
    }
    Ok(conds)
}

fn parse_coded_url(c: &mut Cursor<'_>) -> DavResult<String> {
    debug_assert_eq!(c.peek(), Some(b'<'));
    c.bump();
    let start = c.i;
    loop {
        match c.bump() {
            Some(b'>') => break,
            Some(b'<') => return Err(DavError::BadRequest("nested '<' in If header".into())),
            Some(_) => {
                if c.i - start > MAX_TOKEN_LEN {
                    return Err(DavError::BadRequest("If token too long".into()));
                }
            }
            None => return Err(DavError::BadRequest("unterminated '<' in If header".into())),
        }
    }
    let raw = &c.b[start..c.i - 1];
    if raw.is_empty() {
        return Err(DavError::BadRequest("empty coded-url".into()));
    }
    std::str::from_utf8(raw)
        .map(|s| s.to_string())
        .map_err(|_| DavError::BadRequest("non-UTF-8 coded-url".into()))
}

fn parse_etag(c: &mut Cursor<'_>) -> DavResult<CondKind> {
    debug_assert_eq!(c.peek(), Some(b'['));
    c.bump();
    c.skip_ws();
    let mut weak = false;
    if c.i + 2 <= c.b.len() && c.b[c.i..c.i + 2].eq_ignore_ascii_case(b"W/") {
        weak = true;
        c.i += 2;
    }
    if c.peek() != Some(b'"') {
        return Err(DavError::BadRequest("entity-tag must be quoted".into()));
    }
    c.bump();
    let start = c.i;
    loop {
        match c.bump() {
            Some(b'"') => break,
            Some(_) => {
                if c.i - start > MAX_TOKEN_LEN {
                    return Err(DavError::BadRequest("entity-tag too long".into()));
                }
            }
            None => return Err(DavError::BadRequest("unterminated entity-tag".into())),
        }
    }
    let value = std::str::from_utf8(&c.b[start..c.i - 1])
        .map_err(|_| DavError::BadRequest("non-UTF-8 entity-tag".into()))?
        .to_string();
    c.skip_ws();
    if c.bump() != Some(b']') {
        return Err(DavError::BadRequest("entity-tag missing ']'".into()));
    }
    Ok(CondKind::ETag { value, weak })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn documented_grammar_cases() {
        // RFC 4918 §10.4.6 examples plus what clients actually send.
        let cases = [
            "(<urn:uuid:181d4fae-7d8c-11d0-a765-00a0c91e6bf2>)",
            "(<urn:uuid:181d4fae-7d8c-11d0-a765-00a0c91e6bf2> [\"I am an ETag\"])",
            "([\"I am an ETag\"])",
            "(Not <urn:uuid:181d4fae-7d8c-11d0-a765-00a0c91e6bf2>)",
            "(Not <urn:uuid:a> <urn:uuid:b>)",
            "</resource1> (<urn:uuid:181d4fae-7d8c-11d0-a765-00a0c91e6bf2> [W/\"A weak ETag\"]) ([\"strong\"])",
            "(<urn:uuid:a>) (Not <urn:uuid:b>)",
            "</a/b> (<urn:uuid:x>) </c/d> (<urn:uuid:y>)",
        ];
        for c in cases {
            IfHeader::parse(c).unwrap_or_else(|e| panic!("{c:?} failed: {e}"));
        }
    }

    #[test]
    fn tagged_list_binds_tag_to_each_list() {
        let h = IfHeader::parse("</a> (<urn:uuid:x>) (<urn:uuid:y>)").unwrap();
        assert_eq!(h.lists.len(), 2);
        assert_eq!(h.lists[0].tag.as_deref(), Some("/a"));
        assert_eq!(h.lists[1].tag.as_deref(), Some("/a"));
    }

    #[test]
    fn not_is_case_insensitive_and_not_greedy() {
        let h = IfHeader::parse("(NOT <urn:uuid:x>)").unwrap();
        assert!(h.lists[0].conditions[0].not);
        // "Notmuch" is not a keyword; it is also not a valid condition start.
        assert!(IfHeader::parse("(Notmuch <urn:uuid:x>)").is_err());
    }

    #[test]
    fn malformed_is_rejected() {
        for bad in [
            "",
            "   ",
            "()",
            "(",
            "(<urn:uuid:x>",
            "<urn:uuid:x>",
            "</a>",
            "[\"etag\"]",
            "(<>)",
            "(<urn:uuid:x)",
            "([etag])",
            "([\"etag\")",
            "garbage",
            "(Not)",
        ] {
            assert!(
                IfHeader::parse(bad).is_err(),
                "{bad:?} should not have parsed"
            );
        }
    }

    #[test]
    fn evaluation_or_of_lists_and_of_conditions() {
        let h = IfHeader::parse("(<urn:uuid:x> [\"e1\"]) (<urn:uuid:z>)").unwrap();
        let st = |_: Option<&str>| ResourceState {
            tokens: vec!["urn:uuid:z".into()],
            etag: Some("e9".into()),
            exists: true,
        };
        assert!(h.evaluate(st));

        let st2 = |_: Option<&str>| ResourceState {
            tokens: vec!["urn:uuid:x".into()],
            etag: Some("e9".into()),
            exists: true,
        };
        assert!(!h.evaluate(st2)); // first list fails on etag, second on token
    }

    #[test]
    fn negated_condition_matches_when_absent() {
        let h = IfHeader::parse("(Not <urn:uuid:nope>)").unwrap();
        assert!(h.evaluate(|_| ResourceState {
            tokens: vec![],
            etag: None,
            exists: true
        }));
    }
}

File: go/internal/dav/ifheader.go
package dav

import (
	"fmt"
	"strings"

	"github.com/heavycaffeiner/stowcloud/go/internal/limits"
)

// The If header, RFC 4918 section 10.4.
//
//	If           = 1*( No-tag-list | Tagged-list )
//	No-tag-list  = List
//	Tagged-list  = Resource-Tag 1*List
//	List         = "(" 1*Condition ")"
//	Condition    = ["Not"] ( State-token | "[" entity-tag "]" )
//	State-token  = Coded-URL
//	Coded-URL    = "<" absolute-URI ">"
//	Resource-Tag = "<" Simple-ref ">"
//
// The grammar is only locally ambiguous: at the top level a "<" starts a
// resource tag, inside parentheses it starts a state token. The parser is
// built around that one fact.
//
// Getting this wrong means either honouring a lock nobody holds or ignoring
// one somebody does, which is why it is small, hand-written and fuzzed.

// CondKind is what a condition tests.
type CondKind int

const (
	CondStateToken CondKind = iota
	CondETag
)

// Condition is one term inside a list.
type Condition struct {
	Not   bool
	Kind  CondKind
	Value string
	// Weak records that an entity tag arrived with the W/ marker. A weak
	// validator can never match here, which is what makes it worth recording
	// rather than discarding.
	Weak bool
}

// IfList is one parenthesised list and the resource it was tagged for. Tag is
// empty for a no-tag list, meaning the request URI.
type IfList struct {
	Tag        string
	Tagged     bool
	Conditions []Condition
}

// IfHeader is a parsed If header.
type IfHeader struct {
	Lists []IfList
}

// ResourceState is what the server knows about one resource, for evaluation.
type ResourceState struct {
	// Tokens are the live lock tokens on the resource, in urn:uuid: form.
	Tokens []string
	// ETag is the current validator, unquoted.
	ETag string
	// Weak reports that the validator is weak, which is what every file
	// validator on Linux is here: statx exposes no inode change version, so a
	// metadata-derived token cannot promise the bytes are unchanged.
	Weak bool
	// Exists is false for a resource that is not there, and then only a
	// negated condition can hold.
	Exists bool
}

// ParseIf reads an If header.
func ParseIf(s string) (IfHeader, error) {
	c := &ifCursor{b: []byte(s)}
	var lists []IfList

	c.skipSpace()
	if c.eof() {
		return IfHeader{}, fmt.Errorf("%w: an empty If header", ErrBadRequest)
	}

	for !c.eof() {
		if len(lists) >= limits.DavIfLists {
			return IfHeader{}, limits.Exceed("dav If lists", limits.DavIfLists, int64(len(lists))+1)
		}
		switch c.peek() {
		case '(':
			conds, err := parseIfList(c)
			if err != nil {
				return IfHeader{}, err
			}
			lists = append(lists, IfList{Conditions: conds})

		case '<':
			tag, err := parseCodedURL(c)
			if err != nil {
				return IfHeader{}, err
			}
			c.skipSpace()
			// A tagged list needs at least one list after the tag.
			if c.peek() != '(' {
				return IfHeader{}, fmt.Errorf("%w: a resource tag with no list after it", ErrBadRequest)
			}
			for c.peek() == '(' {
				if len(lists) >= limits.DavIfLists {
					return IfHeader{}, limits.Exceed("dav If lists",
						limits.DavIfLists, int64(len(lists))+1)
				}
				conds, err := parseIfList(c)
				if err != nil {
					return IfHeader{}, err
				}
				lists = append(lists, IfList{Tag: tag, Tagged: true, Conditions: conds})
				c.skipSpace()
			}

		default:
			return IfHeader{}, fmt.Errorf("%w: unexpected %q in an If header",
				ErrBadRequest, string(c.peek()))
		}
		c.skipSpace()
	}

	if len(lists) == 0 {
		return IfHeader{}, fmt.Errorf("%w: an If header with no lists", ErrBadRequest)
	}
	return IfHeader{Lists: lists}, nil
}

// Tokens is every state token the header asserts, ignoring negated ones.
//
// A negated token is an assertion that the client does *not* hold it, so
// treating it as submitted is how a write slips past a lock it never had.
func (h IfHeader) Tokens() []string {
	var out []string
	for _, l := range h.Lists {
		for _, c := range l.Conditions {
			if c.Kind == CondStateToken && !c.Not {
				out = append(out, c.Value)
			}
		}
	}
	return out
}

// Evaluate applies the header. Lists are OR-ed and the conditions inside one
// list are AND-ed.
//
// stateOf is asked for a tagged resource by URL, or for the request URI when
// tagged is false.
func (h IfHeader) Evaluate(stateOf func(tag string, tagged bool) ResourceState) bool {
	for _, list := range h.Lists {
		st := stateOf(list.Tag, list.Tagged)
		all := true
		for _, cond := range list.Conditions {
			hit := false
			switch cond.Kind {
			case CondStateToken:
				for _, t := range st.Tokens {
					if t == cond.Value {
						hit = true
						break
					}
				}
			case CondETag:
				// The tags have to be equal and agree on strength. That is
				// the strong comparison RFC 4918 asks for, applied to what
				// this server actually issues.
				//
				// Refusing every weak tag outright, which this did, made the
				// header useless on this build: every file validator here is
				// weak, so a client echoing back the exact ETag the server
				// had just given it was told the precondition failed. A
				// client that guards writes with If could not write at all.
				//
				// What the weak marker costs is still enforced, because it
				// is the same token either way: two different byte sequences
				// can share this validator only if their size, mtime, ctime
				// and identity all match, and a client is told the guarantee
				// is advisory by the marker itself.
				hit = st.Exists && st.Weak == cond.Weak && st.ETag == cond.Value
			}
			if hit == cond.Not {
				all = false
				break
			}
		}
		if all {
			return true
		}
	}
	return false
}

type ifCursor struct {
	b []byte
	i int
}

func (c *ifCursor) eof() bool { return c.i >= len(c.b) }

func (c *ifCursor) peek() byte {
	if c.eof() {
		return 0
	}
	return c.b[c.i]
}

func (c *ifCursor) bump() byte {
	ch := c.peek()
	if !c.eof() {
		c.i++
	}
	return ch
}

func (c *ifCursor) skipSpace() {
	for !c.eof() {
		switch c.b[c.i] {
		case ' ', '\t', '\r', '\n':
			c.i++
		default:
			return
		}
	}
}

// eatKeyword matches a literal case-insensitively and refuses to swallow the
// front of a longer word, so "Nothing" does not parse as "Not" plus "hing".
func (c *ifCursor) eatKeyword(kw string) bool {
	n := len(kw)
	if c.i+n > len(c.b) {
		return false
	}
	if !strings.EqualFold(string(c.b[c.i:c.i+n]), kw) {
		return false
	}
	if c.i+n < len(c.b) {
		switch c.b[c.i+n] {
		case ' ', '\t', '\r', '\n', '<', '[', '(':
		default:
			return false
		}
	}
	c.i += n
	return true
}

func parseIfList(c *ifCursor) ([]Condition, error) {
	c.bump() // the opening parenthesis
	var conds []Condition

	for {
		c.skipSpace()
		if c.eof() {
			return nil, fmt.Errorf("%w: an unterminated If list", ErrBadRequest)
		}
		if c.peek() == ')' {
			c.bump()
			break
		}
		if len(conds) >= limits.DavIfConditions {
			return nil, limits.Exceed("dav If conditions",
				limits.DavIfConditions, int64(len(conds))+1)
		}

		not := c.eatKeyword("Not")
		c.skipSpace()

		switch c.peek() {
		case '<':
			tok, err := parseCodedURL(c)
			if err != nil {
				return nil, err
			}
			conds = append(conds, Condition{Not: not, Kind: CondStateToken, Value: tok})
		case '[':
			tag, weak, err := parseEntityTag(c)
			if err != nil {
				return nil, err
			}
			conds = append(conds, Condition{Not: not, Kind: CondETag, Value: tag, Weak: weak})
		default:
			return nil, fmt.Errorf("%w: expected a state token or an entity tag in an If list",
				ErrBadRequest)
		}
	}

	if len(conds) == 0 {
		return nil, fmt.Errorf("%w: an empty If list", ErrBadRequest)
	}
	return conds, nil
}

func parseCodedURL(c *ifCursor) (string, error) {
	c.bump() // "<"
	start := c.i
	for !c.eof() && c.peek() != '>' {
		if c.i-start >= limits.DavIfTokenLength {
			return "", limits.Exceed("dav If token", limits.DavIfTokenLength, int64(c.i-start)+1)
		}
		c.bump()
	}
	if c.eof() {
		return "", fmt.Errorf("%w: an unterminated coded URL", ErrBadRequest)
	}
	s := string(c.b[start:c.i])
	c.bump() // ">"
	if s == "" {
		return "", fmt.Errorf("%w: an empty coded URL", ErrBadRequest)
	}
	return s, nil
}

func parseEntityTag(c *ifCursor) (tag string, weak bool, err error) {
	c.bump() // "["
	c.skipSpace()

	if c.i+2 <= len(c.b) && (c.b[c.i] == 'W' || c.b[c.i] == 'w') && c.b[c.i+1] == '/' {
		weak = true
		c.i += 2
	}
	if c.peek() != '"' {
		return "", false, fmt.Errorf("%w: an entity tag that is not quoted", ErrBadRequest)
	}
	c.bump() // the opening quote

	start := c.i
	for !c.eof() && c.peek() != '"' {
		if c.i-start >= limits.DavIfTokenLength {
			return "", false, limits.Exceed("dav If token",
				limits.DavIfTokenLength, int64(c.i-start)+1)
		}
		c.bump()
	}
	if c.eof() {
		return "", false, fmt.Errorf("%w: an unterminated entity tag", ErrBadRequest)
	}
	tag = string(c.b[start:c.i])
	c.bump() // the closing quote

	c.skipSpace()
	if c.peek() != ']' {
		return "", false, fmt.Errorf("%w: an entity tag with no closing bracket", ErrBadRequest)
	}
	c.bump()
	return tag, weak, nil
}
